src/App.js

```diff
@@ -29,37 +30,90 @@ const App = () => {
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
-
+  const [currentPlayer, setCurrentPlayer] = useState(player1);
   // Wave 2
   // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
+  };
+
+  const changeSquares = (id) => {
+    const newSquare = [...squares];
+    console.log(newSquare);
+    for (let row of newSquare) {
+      for (let col of row) {
+        if (checkForWinner() === null) {
+          if (col.id === id && col.value === '') {
+            console.log(col.value);
+            col.value = currentPlayer;
+            console.log(col.value);
+          }
+        } else {
+          break;
+        }
+      }
+      if (currentPlayer === player1) {
+        setCurrentPlayer(player2);
+      } else {
+        setCurrentPlayer(player1);
+      }
+      setSquares(newSquare);
+    }
   };
 
   const resetGame = () => {
-    // Complete in Wave 4
+    setSquares(generateSquares);
   };
 
   return (
     <div className="App">
       <header className="App-header">
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2>Winner is {checkForWinner()}</h2>
```

Comment:
Nice job observing that checkForWinner should be called while calculating the next render pass, instead of during the click handler. Since checkForWinner looks at the squares piece of state, it shouldn't see a change until the next render (though there's a bit of cross contamination happening in your click handler). But what we might additionally do is call checkForWinner before the return, storing the result in a local variable, and then use its result to determine whether to display this section at all (avoid displaying "Winner is" without a winner) using conditional rendering.